Add app shortcuts blog post (#2879)
GoogleChrome/web.dev: src/site/content/en/blog/app-shortcuts/index.md

---
title: Get things done quickly with app shortcuts
subhead: App shortcuts give quick access to a handful of common actions that users need frequently.
authors:
  - beaufortfrancois
  - jungkees
date: 2020-05-27
updated: 2020-05-27
hero: hero.jpg
thumbnail: thumbnail.jpg
alt: A photo of an Android phone showing an app shortcuts menu
description: App shortcuts give quick access to a handful of common actions that users need frequently.
tags:
  - blog # blog is a required tag for the article to show up in the blog.

>   - capabilities

Reviewer: ```suggestion
  - capabilities
  - fugu
```

---
title: Get things done quickly with app shortcuts

> subhead: App shortcuts give quick access to a handful of common actions that users need frequently.

Reviewer 2: Nice summary

authors:
  - beaufortfrancois
  - jungkees
date: 2020-05-27
updated: 2020-05-27

> hero: hero.jpg

Reviewer 2: Weirdly enough this looks OK as a hero but a little grainy as a thumbnail:

![image](https://user-images.githubusercontent.com/4713486/82085479-f60a6980-96a1-11ea-8433-c64da6f98f90.png)

Maybe we can use a less-optimized version of the image?

Author: I realized we don't need a thumbnail at all if using the same image. So I've removed it.

> date: 2020-05-28
> updated: 2020-05-28

Reviewer: ```suggestion
date: 2020-05-27
updated: 2020-05-27
```

thumbnail: thumbnail.jpg
alt: A photo of an Android phone showing an app shortcuts menu
description: App shortcuts give quick access to a handful of common actions that users need frequently.
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - capabilities
  - fugu
  - progressive-web-apps
---

> To improve users' productivity and facilitate re-engagement with key tasks, the
> web platform now supports app shortcuts. They allow web developers to provide
> quick access to a handful of common actions that users need frequently.

Reviewer 2: ```suggestion
To improve users' productivity and facilitate re-engagement with key tasks,
Chromium-based browsers now supports app shortcuts. They allow web developers to provide
quick access to a handful of common actions that users need frequently.
```
Sorry but we can't say this unless Firefox and Safari also support it. If they do then you can disregard this comment. We have specifically had complaints about us implying that Chromium-based browsers are the web platform.

Reviewer 2: It would also be nice to link to an MDN browser compatibility table

Author: Because manifest are part of the manifest spec, which is supported on all browsers I think it's ok in this case to say this. We're not introducing a brand new API, it's an addition to the manifest spec that is widely deployed.

Note that Chromium-based browsers isn't very user friendly.

Author: For info, my original writing before copy-editing was "app shortcuts are added to the web platform". It was changed to "the web platform now supports app shortcuts" as using active voice is a best practice.

Reviewer 2: Pinged you and @petele on Chat to resolve.

Reviewer 2: Noting for the record that I expressed my concerns to Pete and Francois over Chat but they still feel that "web platform" is an appropriate phrase here so we're keeping it.

{% Aside %}
At the time of writing, app shortcuts are available in Chrome 84 for Android. On
Windows, you can experiment with them on both Chrome 84 and Edge 84 by enabling
the flag `about://flags/#enable-desktop-pwas-app-icon-shortcuts-menu`.
{% endAside %}

This article will teach you how to define those app shortcuts. Additionally,
you'll learn some associated best practices.

## About app shortcuts

App shortcuts help users quickly start common or recommended tasks within your
web app. Easy access to those tasks from anywhere the app icon is displayed will
enhance users' productivity as well as increase their engagement with the web
app.

The app shortcuts menu is invoked by right-clicking the app icon in the taskbar
(Windows) or dock (macOS) on the user's desktop, or long pressing the app's
launcher icon on Android.

<figure class="w-figure">

>   <img src="./app-shortcuts-menu-android.png" alt="Screenshot of an app shortcuts menu opened on Android">

Reviewer 2: ```suggestion
  <img class="w-screenshot" src="./app-shortcuts-menu-android.png" alt="Screenshot of an app shortcuts menu opened on Android">
```

Reviewer 2: I *think* this will make the image look a bit nicer. Could be wrong.

  <figcaption class="w-figcaption">App shortcuts menu opened on Android</figcaption>
</figure>

The app shortcuts menu is shown only for [Progressive Web Apps] that are

> installed on the user's desktop or mobile device.

Reviewer 2: Would be nice to explicitly define what criteria are relevant here. Does it have to meet the entire base PWA checklist? Or is it just the existence of the app on the homescreen? Specific properties in the manifest?

Author: PWA have to be installed to the homescreen, that implies all the stuff covered in https://web.dev/install-criteria/. Shall I link to it this article instead?

Reviewer: Say what you just said and link to the article.

Reviewer: Don't hold up the article, though. I'd be fine with submitting this change as an update.

Author: 6b8201d addresses it

      "url": "/subscriptions?utm_source=homescreen",
      "icons": [{ "src": "/icons/subscriptions.png", "sizes": "192x192" }]
    }
  ]
}
```

Each member of the `shortcuts` array is a dictionary that contains at least a
`name` and a `url`. Other members are optional.

### name

The human-readable label for the app shortcut when
displayed to the user.

### short_name (optional)

The human-readable label used where space is limited. It is recommended
that you provide it, even though it's
optional.

### description (optional)

The human-readable purpose for the app shortcut.
It is not used at the time of writing.

### url

The URL opened when a user activates the app
shortcut. This URL must exist within the scope of the web app manifest. If it is
a relative URL, the base URL will be the URL of the web app manifest.

### icons (optional)

An array of image resource objects. Each object must
include the `src` and a `sizes` property. The `type` of image is optional.

SVG files are not supported at the time of writing, use PNG instead.

If you want pixel-perfect icons, provide them in
increments of 48dp. Otherwise, it is recommended that you use a 192x192 pixels

> icon.

Reviewer 2: My $0.02: I'm not sure why we're rehashing general web app manifest concepts here when we have a guide dedicated to the topic. We're also introducing maintainability risk (e.g. we update /add-manifest but forget to update this guide). Are these the properties that are required in order to make app shortcuts work?

Author: Icons for app shortcut icons have different rules than icons for manifest. Hence this description. Note that add-manifest doesn't have this text on purpose.

Reviewer 2: It seems like this post should explicitly mention that app shortcut icons have different rules than manifest icons?

Author: Done in latest patch

Reviewer 3: > It seems like this post should explicitly mention that app shortcut icons have different rules than manifest icons?

How do you mean?

## Test your app shortcuts

To verify your app shortcuts are setup correctly, use the **Manifest** pane in the

> **Application** panel of DevTools.

Reviewer 2: Oh, cool!

Also, I missed this new feature in What's New In DevTools (Chrome 84). Do you have cycles to add a section to the post? It's OK if not, but I just figured it doesn't hurt to ask.

Author: I just added it two days ago ;)
https://chromiumdash.appspot.com/commit/9fee874360dbc0f891609a8aa1a9f8947b510253
Can you link to the post?

Reviewer 2: https://developers.google.com/web/updates/2020/05/devtools/

Author: https://github.com/google/WebFundamentals/pull/8716

> If you prefer to adjust your icons for pixel-perfection, provide icons in

Reviewer: The problem with original phrasing is that it sounded like you were talking some standard you're not linking to instead of a value judgement.

> If you prefer to adjust your icons for pixel-perfection, provide icons in

Reviewer: If you keep the original text, you should link to something that explains it.

<figure class="w-figure">
  <img src="./app-shortcuts-devtools.png" alt="Screenshot of app shortcuts in DevTools">
  <figcaption class="w-figcaption">App shortcuts shown in DevTools</figcaption>
</figure>

This pane provides a human-readable version of many of your manifest's
properties, including app shortcuts. It makes it easy to verify that all of the
shortcut icons, if provided, are loading properly.

> {% Aside 'caution' %}

Reviewer: ```suggestion
{% Aside %}
```

DevTools support for app shortcuts is available in Chrome 84 and Edge 84 and later.
{% endAside %}

App shortcuts may not be available right away to all users because PWA updates
are capped to once a day.  You can find more details in this [article].

## Best practises

### Order app shortcuts by priority

You are encouraged to order app shortcuts by priority, with the most critical
app shortcuts appearing first in the `shortcuts` array as the limit on the
number of app shortcuts displayed varies depending on the platform. Chrome and
Edge on Windows for instance limit the number of app shortcuts to 10 while
Chrome for Android only takes the first 4 app shortcuts into account.

### Use distinct app shortcut names

You should not rely on icons to differentiate app shortcuts as they may not
always be visible. For instance, macOS doesn't support icons in the dock
shortcuts menu. Use distinct names for each app shortcut.

### Measure app shortcuts usage

You should annotate app shortcuts `url` entries like you would do with

> `start_url` for analytics purposes (.e.g. `url:

Reviewer 2: ```suggestion
`start_url` for analytics purposes (e.g. `url:
```

"/my-shortcut?utm_source=homescreen"`).

## Browser support

App shortcuts are available in Chrome 84 on Android. On Windows, you can
experiment with them on both Chrome 84 and Edge 84 by enabling the flag
`about://flags/#enable-desktop-pwas-app-icon-shortcuts-menu`. More desktop

> platform support will follow.

Reviewer 2: I would much prefer this to be the first section of the guide. I know some of our contributors feel that it messes up the flow of the content, but we know from the MDN DNA results that cross-browser compatibility is web developers' number 1 frustration so I view it as a significant show of respect to them to put this information upfront. In other words, for some (most?) developers, when they realize that it's Chromium-only, they will stop reading. Therefore putting it upfront respects the reader's time.

Author: It is already at the very top of the article on purpose.
![image](https://user-images.githubusercontent.com/634478/82090332-a9e91480-96f5-11ea-84ac-1b50b5fcebed.png)

Reviewer 2: Fair enough, thanks.

## Trusted Web Activity support

[Bubblewrap], the recommended tool to build Android apps that use [Trusted Web
Activity], reads app shortcuts from the web app manifest and automatically
generates the corresponding configuration for the Android app.

For developers integrating Trusted Web Activity manually into their Android
application, [Android app shortcuts] can be used to implement the same
behaviors.

## Sample

> Check out the app shortcuts [sample] and its [source].

Reviewer 2: Why not embed the Glitch into this post? 

https://web.dev/handbook/web-dev-components/#glitches

Author: Because it is not actionable. Meaning user has to install the app, and there's not much to see to be honest.

Reviewer 2: But they can immediately browse the source code of the Glitch?

Author: I've added a glitch embed in latest patch.

<video
  autoplay
  loop
  muted
  style=" height: 480px; width: 100%"
  poster="https://storage.cloud.google.com/webfundamentals-assets/videos/app-shortcuts-recording.png"
  src="https://storage.cloud.google.com/webfundamentals-assets/videos/app-shortcuts-recording.mp4">
</video>

## Helpful links

* [Explainer][explainer]
* [Spec][spec]
* [App shortcuts sample][sample] | [App shortcuts sample source][source]
* [Tracking bug][cr-bug]
* [ChromeStatus.com entry][cr-status]
* Blink Component: [`UI>Browser>WebAppInstalls`]

> <!-- lint disable definition-case -->

Reviewer 2: @robdodson why do we enforce so many rules on link IDs? Seems kinda obtrusive IMO. I'm pretty sure that 11ty will barf errors if any of the link IDs don't work, so I'm not sure why we're enforcing specific conventions.

Reviewer 4: Our remarklintrc file uses the [remark-preset-lint-markdown-style-guide](https://github.com/remarkjs/remark-lint/tree/master/packages/remark-preset-lint-markdown-style-guide#rules) as a way to set a number of helpful defaults (https://github.com/GoogleChrome/web.dev/blob/master/.remarkrc.js#L8). We used this preset because we didn't know what rules we would want, so it's easier to start with a preset and disable things that you find annoying.

You can turn off this rule globally by adding `['lint-definition-case', false],` to that file.
